website/src/components/Submission/FileUpload/ColumnMapping.ts

```diff
@@ -0,0 +1,99 @@
+        const [bestMatch, score] = inputFields
+            .map((field): [string, number] => {
+                const score = Math.max(
+                    stringSimilarity(sourceColumn, field.name),
+                    stringSimilarity(sourceColumn, field.displayName ?? ''),
+                );
+                return [field.name, score];
+            })
+            .reduce((maxItem, currentItem) => (currentItem[1] > maxItem[1] ? currentItem : maxItem));
+        return score > 0.8 ? bestMatch : null;
+    }
+
+    /* Create a new mapping with the given columns, doing a best-effort to pre-match columns. */
+    public static fromColumns(sourceColumns: string[], inputFields: InputField[]) {
+        const mapping = new Map();
+        let availableFields = inputFields;
+        sourceColumns.forEach((sourceColumn) => {
+            const bestMatch = this.getBestMatchingTargetColumn(sourceColumn, availableFields);
+            mapping.set(sourceColumn, bestMatch);
+            availableFields = availableFields.filter((field) => field.name !== bestMatch);
+        });
+        return new ColumnMapping(mapping);
+    }
+
+    /* Update the mapping with new source and target columns, trying to keep as much of the 
+       mapping intact as possible. */
+    public update(newSourceColumns: string[], inputFields: InputField[]): ColumnMapping {
+        const newMapping = new Map(
+            newSourceColumns.map((newSourceCol) => {
+                const prevTargetCol = this.map.get(newSourceCol);
+                if (prevTargetCol && inputFields.map((f) => f.name).includes(prevTargetCol)) {
+                    return [newSourceCol, prevTargetCol];
+                } else {
+                    return [newSourceCol, ColumnMapping.getBestMatchingTargetColumn(newSourceCol, inputFields)];
+                }
+            }),
+        );
+        return new ColumnMapping(newMapping);
+    }
+
+    /* Returns the entries in the mapping as a list. Each item in the list has:
+     * - The source column name
+     * - The target column name
+     */
+    public entries(): [string, string | null][] {
+        return Array.from(this.map.entries());
+    }
+
+    public usedColumns(): string[] {
+        return Array.from(this.map.values()).filter((v): v is string => v !== null);
+    }
+
+    public updateWith(sourceColumn: string, targetColumn: string | null): ColumnMapping {
+        const newMapping = new Map(this.map);
+        newMapping.set(sourceColumn, targetColumn);
+        this.map.forEach((targetCol, srcCol) => targetCol === targetColumn && newMapping.set(srcCol, null));
+        return new ColumnMapping(newMapping);
+    }
+
+    /* Apply this mapping to a TSV file, returning a new file with remapped columns. */
+    public async applyTo(tsvFile: ProcessedFile): Promise<File> {
+        const text = await tsvFile.text();
+        const inputRows = text.trim().split('\n');
```

Comment:
Oh, you're not using a full TSV parser? This would cause problems if there is a multiline cell which, I think, is a reasonable use case for fields like version comment. The following file doesn't work:

```
submissionId	versionComment	geoLocCountry	sampleCollectionDate
test1	"Paragraph 1...

...and paragraph 2"	USA	2011-07-18
test2	"Paragraph 1...

...and paragraph 2"	USA	2011-07-18
```

Reply:
update: used a different lib.